### DISCUSSION TITLE
[SOLVED] Idea : Reader Mode

### DISCUSSION THREAD

**Discussion by user858753257** [martinrotter/rssguard#559] [Ideas]
Wouldn't it make sense to include in the internal browser a reading mode like chrome/safari ? 
Or a general reader mode for all articles

**Comment by martinrotter** [comment]
Hello. Well, why not. You mean something like [readablity](https://github.com/mozilla/readability)?

**Reply by martinrotter** [reply]
@F3000 ping

**Comment by akinokonomi** [comment]
Is it like Firefox's built-in Reader view? Which allows to simplify/clean the page's content, and to customize text to some extent (line height, column width, font, colour mode). 

If so, I'm not sure if it would fit in RSS reader, to be honest; in RSS Guard it's the work for "skin".

For it to work properly, it would need keep state between all articles, otherwise it doesn't make sense if user would be obliged to toggle and tune the Reader view for each article.

That is about "a general reader mode for all articles".

As to "include in the internal browser", I think it would make some sense if it can be done.

Here is an example of Reader view for this page https://www.w3schools.com/html/html_editors.asp
![image](https://user-images.githubusercontent.com/91767492/147005772-324b4ed2-1c27-4ae4-b339-d27a606db6dd.png)

Some customization which Firefox allows
![image](https://user-images.githubusercontent.com/91767492/147005836-a125481a-fd66-479a-9147-779f103f8c89.png)

**Comment by martinrotter** [comment]
@F3000 Pls, reply.

One more comment: The "reader mode" is not just about "stylling" of the article. AFAIK Readability.js also has some magic which removes "non-article" parts of website, transforms the actual "article" to be layouted better etc.

When the HTML data is parsed with Readability, it returns pure HTML-formatted text, AFAIK stripped from custom CSS classes/formatting so this data can be for example displayed using style from active skin, and this would make sense to have implemented if I decide to implement "reader" mode.

Also, note there are many "readability" libraries based on Node.js:
* https://github.com/mozilla/readability
* https://github.com/postlight/mercury-parser
* https://www.npmjs.com/package/article-parser
* https://github.com/ibrod83/nodejs-web-scraper

Also, there are many libs written in Python, but I would prefer perhaps Node.js (?) because RSS Guard already uses AdBlock component based on Node.js library, I would like to avoid add dependencies on other runtimes etc.

It seems that mozilla's readability is probably the best OSS solution on the market tho.

**Reply by akinokonomi** [reply]
> One more comment: The "reader mode" is not just about "stylling" of the article. AFAIK Readability.js also has some magic [...]

:+1: Yes, it certainly has.

It is very complex, and would be really difficult to create from the scratch to have it built-in.

> Also, note there are many "readability" libraries based on Node.js

Again, it would be just as entangled to install for user as Adblock, and is likely to be in even less favour because of that.

Now, the question stands whether it is worth it, in the first place, if the process of installation would look so uncanny.

---

@F3000, As a side note, qt-webengine (the internal browser of RSS Guard), though based on Chromium, is exceedingly stripped, bare module. 
And, as far as I know, does not have anything of the sort built-in. So, having such feature is more demanding than it may seem to be at first.

**Reply by martinrotter** [reply]
> Again, it would be just as entangled to install for user as Adblock, and is likely to be in even less favour because of that.

I already added some code to make usage of node.js stuff in rssguard more easier as global `NODE_PATH` is now deduced automatically by rssguard (as long as `node` and `npm` are installed and are on the `PATH`). I think about making the setup of adblock (and potentially this) even more easy.

**Reply by martinrotter** [reply]
> As a side note, qt-webengine (the internal browser of RSS Guard), though based on Chromium, is exceedingly stripped, bare module.

Not exceedingly, but yes, it is quite bare.

Also, speaking of ease of installation. I think I could even automagically install those needed node modules directly in rssguard, into "local" node-modules path (preferably under user-data folder). That would only require user to actually install node/npm which is easy on Linux and on Windows where you just run next-next-next installer.

Other possibility would be to drop node/npm dependency and use python, it offers adblock component too and also "readability" stuff.

**Reply by akinokonomi** [reply]
> I think about making the setup of adblock (and potentially this) even more easy.

~~Even more?~~ (You answered in your next comment.)

**Reply by akinokonomi** [reply]
> install those needed node modules directly in rssguard, into "local" node-modules path (preferably under user-data folder)

I hope you mean local node directory, not RSS Guard's config/local folders?

> on Windows where you just run next-next-next installer

Hahaha.

But it needs to be up-to-date, right. From user's point of view it may look as something demanding, if they does not use node/npm themself.
They would just install and forget about it, and it will always be outdated, isn't it? which is not a nice thing.

**Reply by martinrotter** [reply]
Btw to elaborate more. Node vs Python - not really sure which one is "better". I just know Python bit more and is used in example scraper scripts in repository.

If i should make installation of these things easier, it is also more or less equal, both node and python provide ways of installing packages  "locally" without sudo/admin rights etc. It just seems that python as more "strongly" typed than node is sometimes easier to write and read....

**Reply by martinrotter** [reply]
> I hope you mean local node directory, not RSS Guard's config/local folders?

You have either `$(npm root -g)` which is like `/usr/something/something` on Linux. Or you can have your node modules basically anywhere you want, for example in `/home/user/.config/rssguard/node-modules` etc.

I would suggest to have modules in RSS GUard user data folder, in `modules-win`, `modules-linux`, `modules-osx` folders, decided in runtime. That way, modules will be transferable (as many users sync their rss guard user data folders). Also node modules are not generally cross-platform, that's why I suggest having OS name in the folder.

**Reply by martinrotter** [reply]
> But it needs to be up-to-date, right. From user's point of view it may look as something demanding, if they does not use node/npm themself.
> They would just install and forget about it, and it will always be outdated, isn't it? which is not a nice thing.

Well, on Linux, we do not suffer from this right. I run my `pacman` regularly and I update those `pip` and `npm` modules once in a while too. On Windows, this could be problem in theory, but fact is that those modules do not really break that much. It really depends on how they will be used in RSS Guard and I prefer to be more conservative. I wanted to avoid having any dependencies, but C++ AdBlock component is hard to write from scratch and previous solution was just incredibly bad/slow.

**Reply by akinokonomi** [reply]
> I would suggest to have modules in RSS GUard user data folder, [...]

Hmm. I guess it could work.

Sorry, I cannot be of much help here. The only things I currently use npm for is to build some projects (I use nvm, by the way), as for something other, it was quite a time ago and I don't remember a thing...

Is there some project that does it this way?

**Reply by martinrotter** [reply]
Yes, there are projects that use node with locally installed packages, actually this is more like "standard" use. But most of those projects are like web "apps". For example, AFAIK "Electron" apps even "bundle" node packages and runtime into single "executable".

It is even recommended to have packages local for number of reasons: lack of admin rights, packages do not conflict with global ones, each app requires different version of packages etc.

I will try to implement behavior and use it with adblock and test it if it works and really eases the actual "installation".

**Reply by akinokonomi** [reply]
> But it needs to be up-to-date, right. [...]

Well, the alternative of RSS Guard checking for npm module updates, too, and prom**p**ting user to do so, sounds a bit odd, but it can be the way?

**Reply by akinokonomi** [reply]
> But most of those projects are like web "apps".

Exactly, that's what I also thought, but what about others?

> For example, AFAIK "Electron" apps even "bundle" node packages and runtime into single "executable"

(I use npm pretty much only for Electron indeed.)

**Reply by martinrotter** [reply]
> > But it needs to be up-to-date, right. [...]
> 
> Well, the alternative of RSS Guard checking for npm module updates, too, and promoting user to do so, sounds a bit odd, but it can be the way?

That is certainly the way.

**Reply by martinrotter** [reply]
> Exactly, that's what I also thought, but what about others?

Did not look really. Not sure there are others.

**Reply by akinokonomi** [reply]
Why did you chose node.js over python?

**Reply by martinrotter** [reply]
It was because I was looking for something to replace old C++ adblock implementation with. The only good solution I found was this: https://github.com/ghostery/adblocker/blob/master/packages/adblocker/README.md

Also, there is probably even better nodejs solution: https://github.com/brave/adblock-rust

I did not find any python-based adblock libraries then. Looking now, I see some solutions, I probably just missed them: https://github.com/ArniDagur/python-adblock

So looking at it now, it is actually hard to decide whether node or python is better to integrate...

**Reply by akinokonomi** [reply]
> I would suggest to have modules in RSS GUard user data folder,

Yes, I remembered the way node does things, and this is certainly preferable to installing those packages globally.

> > ... sounds a bit odd, but it can be the way?
>
> That is certainly the way.

And together with this user will just let RSS Guard deal with them. Sounds nice.

**Reply by martinrotter** [reply]
Yes, I will do that with Adblock to just see how it works in real and we can go from there.

**Comment by martinrotter** [comment]
Btw: Here is simple node.js script to run readable on HTML content.
```node
var rd = require('@mozilla/readability');
var fs = require('fs');
var jsd = require('jsdom');
var iconvlite = require('iconv-lite');

//var mrc = require('@postlight/mercury-parser');

var data = fs.readFileSync(0);
var encData = iconvlite.decode(data, "win-1250"); // utf-8 should be used, win-1250 just for my testing
var doc = new jsd.JSDOM(encData);

var article = new rd.Readability(doc.window.document).parse();
console.log(article['content']);
```

Can be called like `curl "website" | node "script.js"`

Something like this would essentially work on RSS Guard with not great amount of coding.

**Reply by akinokonomi** [reply]
Oh, indeed! I remember now. (Used to do some few things with puppeteer myself at some point.)

Just tried. Works great, looks fun!

**Comment by martinrotter** [comment]
@akinokonomi 

> If so, I'm not sure if it would fit in RSS reader, to be honest; in RSS Guard it's the work for "skin".

Yes. When article from feed is displayed, then yes, formatting is just better to be done by skin as it stands now, but when user clicks "URL" or just loads any other website in internal browser, then "readability.js" could be indeed useful, it could be even subsequently styled with skin's css. I still has this in my mind, did not forget about it.

**Comment by martinrotter** [comment]
OK, I reimplemented Node.js integration in RSS Guard which is now de-facto automatic. If node/npm binaries are detected (on Linux they are installed to `PATH`, on Windows the installer also adds them to `PATH` by default), then RSS Guard automatically installs needed package for the functionality when it is first used.

At this point, the only opt-in node.js package needed is `@cliqz/adblocker` for adblock. There is new settings category to tweak node.js.

![image](https://user-images.githubusercontent.com/1255302/153597010-d93b7ea1-2c3c-4d0f-aaa8-6bd17f6491c0.png)

RSS Guard now by default installs packages in a local scope, to a OS-derived subfolder in RSS Guard's `data4` user-data folder. This means no admin rights are needed and installed packages are nicely contained in a folder which is easy to by synced etc.

**Reply by martinrotter** [reply]
Initial tests:

![image](https://user-images.githubusercontent.com/1255302/153605101-acf40d04-7f26-4863-b31f-4c597fe2832d.png)

**Reply by martinrotter** [reply]
Before

![image](https://user-images.githubusercontent.com/1255302/153605593-2f7b62bb-3506-4447-8a5a-852973adef0e.png)

After

![image](https://user-images.githubusercontent.com/1255302/153605636-82affd66-03b5-44a4-b183-5133a9609646.png)

![image](https://user-images.githubusercontent.com/1255302/153605653-262d4156-45e5-4108-87c5-f1e1637e9c23.png)
